### PR TITLE
quick and dirty way of checking pre and post stack effect

### PR DESCRIPTION
**Pull request by Z-Shang**
As mentioned in [this issue](https://github.com/vstinner/bytecode/issues/68),
I've modified compute stack size to take in consideration of the prerequisite stack size for some ops (e.g. `BUILD_CONST_KEY_MAP`)

For these definitely failing cases:
image

Before:
image

After:
image

### CODE REVIEW

**Comment by MatthieuDartiailh** [comment]
This looks reasonable enough. I will give a more detailed read later and try to remember why NOP used to be a problem.

**Comment by Z-Shang** [comment]
> This looks reasonable enough. I will give a more detailed read later and try to remember why NOP used to be a problem.

Thank you 👍

**Review by MatthieuDartiailh — CHANGES_REQUESTED** [review]
Some quick nitpick I caught through this a second time. Next I need to sit and check that we do catch all opcodes (for all supported Python version). I wish CPython would provide more tooling to support this.

File: bytecode/instr.py [hunk 1]
```diff
@@ -282,6 +282,27 @@ def stack_effect(self, jump=None):
         else:
             return dis.stack_effect(self._opcode, arg, jump=jump)
 
+    def pre_and_post_stack_effect(self, jump=None):
+        def _pushes_back(opname):
```
[the remark this answers is not in the document]
Comment (Z-Shang) [reply bytecode/instr.py:286]: sure, I'll take a look at how the updates in Python 3.9 affect this

File: bytecode/instr.py [hunk 2]
```diff
@@ -282,6 +282,27 @@ def stack_effect(self, jump=None):
...
+                or opname.startswith("GET_")
+                or opname.startswith("BINARY_")
+                or opname.startswith("INPLACE_")
+                or opname.startswith("BUILD_")
+                or opname.startswith("CALL_")
+            )
+
+        _effect = self.stack_effect(jump=jump)
+        # To compute pre size and post size to avoid segfault cause by not enough stack element
+        if _opcode.opname[self._opcode].startswith("DUP_TOP"):
```
[the remark this answers is not in the document]
Comment (Z-Shang) [reply bytecode/instr.py:310]: by changing `pushes_back` to a free function, this will only be used here.

File: bytecode/instr.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Z-Shang) [reply bytecode/instr.py]: I've added the Python 3.9 new ops, but by hard coding doesn't feel right.
We may have to find a better way in the future

File: bytecode/tests/test_concrete.py [hunk 3]
```diff
@@ -1309,7 +1309,8 @@ def test_compute_jumps_convergence(self):
         code = Bytecode()
         label1 = Label()
         label2 = Label()
-        nop = "UNARY_POSITIVE"  # don't use NOP, dis.stack_effect will raise
+        nop = "NOP"  # don't use NOP, dis.stack_effect will raise
+        # Why? Changing to NOP seems to have no harm
```
[the remark this answers is not in the document]
Comment (Z-Shang) [reply bytecode/tests/test_concrete.py]: done

File: bytecode/instr.py [hunk 4]
```diff
@@ -282,6 +282,37 @@ def stack_effect(self, jump=None):
         else:
             return dis.stack_effect(self._opcode, arg, jump=jump)
 
+    def pushes_back(self):
```
Comment (MatthieuDartiailh) [inline bytecode/instr.py:285]: Should have a leading underscore, this is meant to be private. And it does not need to be a method, we could extract the opcode name once, and pass it to an independent function, but this should be fine.

File: bytecode/instr.py [hunk 5]
```diff
@@ -282,6 +282,37 @@ def stack_effect(self, jump=None):
...
+            "DICT_MERGE",
+            "IS_OP",
+            "CONTAINS_OP",
+        ]
+
+    def pre_and_post_stack_effect(self, jump=None):
+
+        _effect = self.stack_effect(jump=jump)
+        # To compute pre size and post size to avoid segfault cause by not enough stack element
+        if _opcode.opname[self._opcode].startswith("DUP_TOP"):
```
Comment (MatthieuDartiailh) [inline bytecode/instr.py:310]: We may want to handle the ROT_ opcodes in the same way since I assume the interpreter crashes if there are not enough objects on the stack.

File: bytecode/tests/test_bytecode.py [hunk 6]
```diff
@@ -253,6 +253,34 @@ def test_to_code(self):
...
+            code.compute_stacksize()
+
+    def test_negative_size_call(self):
+        code = Bytecode()
+        code.first_lineno = 1
+        code.extend([Instr("CALL_FUNCTION", 0)])
+        with self.assertRaises(RuntimeError):
+            code.compute_stacksize()
+
+    def test_negative_size_build(self):
```
Comment (MatthieuDartiailh) [inline bytecode/tests/test_bytecode.py:277]: I would like to see some extra tests in particular for corner cases like DUP_TOP.

**Review by MatthieuDartiailh — CHANGES_REQUESTED** [review]

**Review by MatthieuDartiailh — CHANGES_REQUESTED** [review]
Can you also rebase on master ? I will do my best to give this the thorough read it deserves next week.

File: bytecode/instr.py [hunk 7]
```diff
@@ -34,6 +34,27 @@ def const_key(obj):
...
+        or opname.startswith("CALL_")
+    ) or opname in [
+        "LIST_TO_TUPLE",
+        "LIST_EXTEND",
+        "SET_UPDATE",
+        "DICT_UPDATE",
+        "DICT_MERGE",
+        "IS_OP",
+        "CONTAINS_OP",
+    ]
```
Comment (MatthieuDartiailh) [inline bytecode/instr.py:55]: I believe the following are missing from the list of push back (You could use a tuple rather than a list by the way):
SET_ADD
LIST_APPEND
MAP_ADD
IMPORT_
MAKE_FUNCTION
FORMAT_VALUE

File: bytecode/instr.py [hunk 8]
```diff
@@ -282,6 +303,19 @@ def stack_effect(self, jump=None):
...
+
+        _effect = self.stack_effect(jump=jump)
+        # To compute pre size and post size to avoid segfault cause by not enough stack element
+        _opname = _opcode.opname[self._opcode]
+        if _opname.startswith("DUP_TOP"):
+            return _effect * -1, _effect * 2
+        if _pushes_back(_opname):
+            return _effect - 1, 1
+        return {"ROT_TWO": (-2, 2), "ROT_THREE": (-3, 3), "ROT_FOUR": (-4, 4)}.get(
+            _opname, (_effect, 0)
```
[the remark this answers is not in the document]
Comment (MatthieuDartiailh) [reply bytecode/instr.py:316]: UNPACK_* opcodes need to be handled separately since they consume one value and pushed multiple ones

**Review by MatthieuDartiailh — CHANGES_REQUESTED** [review]
I noticed a bunch of possible issues. Please have a look and try to add additional tests to cover those.

File: bytecode/instr.py (hunk 7 above)
Comment (Z-Shang) [reply bytecode/instr.py:55]: it is quite tricky for `IMPORT_` OPs
as `IMPORT_NAME` pops 2 and pushes 1
`IMPORT_FROM` pops 0 and pushes 1
`IMPORT_STAR` simply pops 1 and pushes nothing.

I think it then makes sense to consider only `IMPORT_NAME`

**Comment by Z-Shang** [comment]
I'm just wondering, it is possible to get the upstream (id est the Python team) to support this as a part of the standard library (such as `dis.stack_effect`)...

File: bytecode/instr.py (hunk 7 above)
Comment (MatthieuDartiailh) [reply bytecode/instr.py]: I agree for IMPORT_

**Comment by MatthieuDartiailh** [comment]
Once we consolidate that here I will try to reach to the Python developers to suggest something along those lines indeed.

**Comment by MatthieuDartiailh** [comment]
Changes look nice. I will give it another thorough read during the week and convince myself it will always work (some exception handling have been worrying me but since the tests are passing I think everything is fine and just need to go through the logic once again myself). Hopefully I will merge once I am done. Could you add an entry to the change log in the meantime ?

**Comment by Z-Shang** [comment]
> Changes look nice. I will give it another thorough read during the week and convince myself it will always work (some exception handling have been worrying me but since the tests are passing I think everything is fine and just need to go through the logic once again myself). Hopefully I will merge once I am done. Could you add an entry to the change log in the meantime ?

for the changelog, do I also have to update the version number?

**Comment by MatthieuDartiailh** [comment]
You can just keep adding below 0.12 since it has not been released.

File: bytecode/instr.py [hunk 9]
```diff
@@ -34,6 +34,37 @@ def const_key(obj):
...
+        "IS_OP",
+        "CONTAINS_OP",
+        "FORMAT_VALUE",
+        "MAKE_FUNCTION",
+        "IMPORT_NAME",
+        # technically, these three do not push back, but leave the container
+        # object on TOS
+        "SET_ADD",
+        "LIST_APPEND",
+        "MAP_ADD",
```
Comment (MatthieuDartiailh) [inline bytecode/instr.py:64]: Going through the list of opcodes again LOAD_ATTR should be listed

File: bytecode/instr.py [hunk 10]
```diff
@@ -282,6 +313,27 @@ def stack_effect(self, jump=None):
         else:
             return dis.stack_effect(self._opcode, arg, jump=jump)
 
+    def pre_and_post_stack_effect(self, jump=None):
+
+        _effect = self.stack_effect(jump=jump)
```
Comment (MatthieuDartiailh) [inline bytecode/instr.py:318]: FOR_ITER needs also a special case since it adds to the stack based on TOS for a jump not taken and pop TOS when jumping

File: bytecode/cfg.py [hunk 11]
```diff
@@ -140,7 +141,7 @@ def update_size(delta, size, maxsize):
         # jump is taken.
         if instr.has_jump():
             taken_size, maxsize = update_size(
-                instr.stack_effect(jump=True), size, maxsize
+                *instr.pre_and_post_stack_effect(jump=True), size, maxsize
```
Comment (MatthieuDartiailh) [inline bytecode/cfg.py:144]: It could be nice while we stabilize this to have a way to bypass this in case we get the computation wrong (using an extra keyword to` _compute_stack_size`)

**Review by MatthieuDartiailh — CHANGES_REQUESTED** [review]
Going through the opcodes again I found a couple of cases I missed. Sorry for that. Also since I feel we may still get it wrong it may be interesting to provide a way to skip this check through an extra keyword argument.

Thanks for your work on this !

File: bytecode/cfg.py (hunk 11 above)
Comment (Z-Shang) [reply bytecode/cfg.py:144]: I've made it a kw only argument :)

File: bytecode/instr.py (hunk 10 above)
Comment (Z-Shang) [reply bytecode/instr.py:318]: actually I don't think this needs special consideration since the stack effect of  `FOR_ITER` is push 1 when no jump and pop 1 when jump, since the stack effect of non-pushing backs are all returned as "pre effect", this should be just fine.

Comment (MatthieuDartiailh) [reply bytecode/instr.py:318]: But in the first case you need something of the stack so for your analysis to hold you would need a pre effect of -1 and a post effect of 2, no ?

Comment (Z-Shang) [reply bytecode/instr.py:318]: ah yea, that makes sense then, thank you for the clarification

File: bytecode/cfg.py [hunk 12]
```diff
@@ -141,7 +141,11 @@ def update_size(pre_delta, post_delta, size, maxsize):
         # jump is taken.
         if instr.has_jump():
             taken_size, maxsize = update_size(
```
Comment (MatthieuDartiailh) [inline bytecode/cfg.py:143]: I would do the following to not have to pass check_pre_and_post to the instruction.
```python
effects = instr.pre_and_post_stack_effect(jump=True) if check_pre_and_post else (0, instr.stack_effect(jump=True)
taken_size, maxsize = update_size(*effects, size, maxsize)
```
and something similar for the jump False case.

File: bytecode/instr.py [hunk 13]
```diff
@@ -313,9 +314,12 @@ def stack_effect(self, jump=None):
         else:
             return dis.stack_effect(self._opcode, arg, jump=jump)
 
-    def pre_and_post_stack_effect(self, jump=None):
-
+    def pre_and_post_stack_effect(self, jump=None, check_pre_and_post=True):
```
Comment (MatthieuDartiailh) [inline bytecode/instr.py:317]: I dislike the idea of this function not doing what its name suggests see my above suggestion.

File: bytecode/tests/test_bytecode.py [hunk 14]
```diff
@@ -253,6 +253,62 @@ def test_to_code(self):
...
```
Comment (MatthieuDartiailh) [inline bytecode/tests/test_bytecode.py:284]: This test could be parametrize for all other similar opcodes.

File: bytecode/instr.py [hunk 15]
```diff
@@ -282,6 +314,33 @@ def stack_effect(self, jump=None):
...
+            # if the op pushes value back to the stack, then the stack effect given by dis.stack_effect
+            # actually equals pre + post effect, therefore we need -1 from the stack effect as a pre
+            # condition
+            return _effect - 1, 1
+        if _opname.startswith("UNPACK_"):
+            # Instr(UNPACK_* , n) pops 1 and pushes n
+            # _effect = n - 1
+            # hence we return -1, _effect + 1
+            return -1, _effect + 1
+        if _opname == "FOR_ITER" and not jump:
```
Comment (MatthieuDartiailh) [inline bytecode/instr.py:337]: Having a test for this case would be nice

File: bytecode/instr.py [hunk 16]
```diff
@@ -282,6 +314,33 @@ def stack_effect(self, jump=None):
...
+        # To compute pre size and post size to avoid segfault cause by not enough stack element
+        _opname = _opcode.opname[self._opcode]
+        if _opname.startswith("DUP_TOP"):
+            return _effect * -1, _effect * 2
+        if _pushes_back(_opname):
+            # if the op pushes value back to the stack, then the stack effect given by dis.stack_effect
+            # actually equals pre + post effect, therefore we need -1 from the stack effect as a pre
+            # condition
+            return _effect - 1, 1
+        if _opname.startswith("UNPACK_"):
```
Comment (MatthieuDartiailh) [inline bytecode/instr.py:332]: Having a test for this case would be nice

File: bytecode/tests/test_bytecode.py [hunk 17]
```diff
@@ -253,6 +253,62 @@ def test_to_code(self):
...
+            code.compute_stacksize()
+
+    def test_not_enough_dup(self):
+        code = Bytecode()
+        code.first_lineno = 1
+        code.extend([Instr("LOAD_CONST", 1), Instr("DUP_TOP_TWO")])
+        with self.assertRaises(RuntimeError):
+            code.compute_stacksize()
+
+    def test_not_enough_rot(self):
```
Comment (MatthieuDartiailh) [inline bytecode/tests/test_bytecode.py:305]: This test could be parametrized too to test all rot.

**Review by MatthieuDartiailh — CHANGES_REQUESTED** [review]
Couple minor things, also tests and documentation could be updated. Given the number of iterations I am putting you through let me know if you want me to help with anything.

File: bytecode/cfg.py (hunk 12 above)
Comment (Z-Shang) [reply bytecode/cfg.py:143]: yea, that was also my first thought, but it seems that the coverage test would fail, that's why I changed it to this way

File: bytecode/instr.py (hunk 15 above)
Comment (Z-Shang) [reply bytecode/instr.py:337]: I think I'd like some help on constructing a test case for this one, it would be great if you can help :)

**Comment by MatthieuDartiailh** [comment]
Sorry I have been awfully busy these past weeks. I will try to add the test you asked me to.

**Comment by Z-Shang** [comment]
No worries, I'm also busy as hell these days :(

**Comment by MatthieuDartiailh** [comment]
This looks really good. I will add a bit to the documentation in the coming week and then merge. Sorry again for the slow process I just moved from the USA back to France and I have been quite busy as a consequence.

**Comment by Z-Shang** [comment]
Pas de souci et prends ton temps :)

**Comment by MatthieuDartiailh** [comment]
I rebased to get tests running on 3.9. Once CIs run I will merge shortly. Thanks again for your contribution !

**Review by MatthieuDartiailh — APPROVED** [review]
